Code/GraphMol/FileParsers/MultithreadedSDMolSupplier.cpp

```diff
@@ -0,0 +1,262 @@
+            if (inStream.eof()) {
+              throw FileParseException("End of data field name not found");
+            }
+          }
+        } else {
+          dlabel = tempStr.substr(sl + 1, se - sl - 1);
+          // we know the label - now read in the relevant properties
+          // until we hit a blank line
+          std::getline(inStream, tempStr);
+
+          std::string prop = "";
+          std::string stmp = strip(tempStr);
+          int nplines = 0;  // number of lines for this property
+          while (stmp.length() != 0 || tempStr[0] == ' ' ||
+                 tempStr[0] == '\t') {
+            nplines++;
+            if (nplines > 1) {
+              prop += "\n";
+            }
+            // take off \r if it's still in the property:
+            if (tempStr[tempStr.length() - 1] == '\r') {
+              tempStr.erase(tempStr.length() - 1);
+            }
+            prop += tempStr;
+            // erase tempStr in case the file does not end with a carrier
+            // return (we will end up in an infinite loop if we don't do
+            // this and we do not check for EOF in this while loop body)
+            tempStr.erase();
+            std::getline(inStream, tempStr);
+            stmp = strip(tempStr);
+          }
+          mol->setProp(dlabel, prop);
+          if (df_processPropertyLists) {
+            // apply this as an atom property list if that's appropriate
+            FileParserUtils::processMolPropertyList(*mol, dlabel);
+          }
+        }
+      } else {
+        if (df_strictParsing) {
+          // at this point we should always be at a line starting with '>'
+          // following a blank line. If this is not true and df_strictParsing
+          // is true, then throw an exception, otherwise truncate the rest of
+          // the data field following the blank line until the next '>' or EOF
+          // and issue a warning
+          // FIX: should we be deleting the molecule (which is probably fine)
+          // because we couldn't read the data ???
+          throw FileParseException("Problems encountered parsing data fields");
+        } else {
+          if (!warningIssued) {
+            if (hasProp) {
+              BOOST_LOG(rdWarningLog)
+                  << "Property <" << dlabel
+                  << "> will be truncated after the first blank line\n";
+            } else {
+              BOOST_LOG(rdWarningLog) << "Spurious data before the first "
+                                         "property will be ignored\n";
+            }
+            warningIssued = true;
+          }
+        }
+      }
+    }
+    std::getline(inStream, tempStr);
+  }
+}
+
+ROMol *MultithreadedSDMolSupplier::processMoleculeRecord(
+    const std::string &record, unsigned int lineNum) {
+  PRECONDITION(dp_inStream, "no stream");
+  std::istringstream inStream(record);
+  auto res = MolDataStreamToMol(inStream, lineNum, df_sanitize, df_removeHs,
```

Comment:
You could also have used an SDMolSupplier on the stream (or raw text) and also get the properties.  This would have reduced some code.

Reply:
@bp-kelley : I think I looked at that and determined that it wouldn't work because we want to have an accurate `lineNum` available. I agree that we should refactor this at some point, but I think that approach doesn't work at the moment.

Reply:
Great.  We can make a feature request to seed sdmolsupplier with a line number in the future.